pkg/destroy/nutanix/nutanix.go

```diff
@@ -0,0 +1,223 @@
+		if hasCategoryAssigned(v.Metadata, ec) {
+			matchedVirtualMachineList = append(matchedVirtualMachineList, v)
+		}
+	}
+
+	if len(matchedVirtualMachineList) == 0 {
+		o.logger.Infof("No VMs found that require deletion for cluster %s", o.clusterID)
+	} else {
+		logToBeDeletedVMs(matchedVirtualMachineList, o.logger)
+		err := deleteVMs(o.v3Client.V3, matchedVirtualMachineList, o.logger)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func cleanupImages(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	bootISOImageName := nutanixtypes.BootISOImageName(o.infraID)
+	rhcosImageName := nutanixtypes.RHCOSImageName(o.infraID)
+	expectedImageNames := []string{
+		bootISOImageName,
+		rhcosImageName,
+	}
+
+	var found bool
+	allImages, err := o.v3Client.V3.ListAllImage(emptyFilter)
+	if err != nil {
+		return err
+	}
+
+	for _, expectedImageName := range expectedImageNames {
+		found = false
+		for _, i := range allImages.Entities {
+			imageName := *i.Spec.Name
+			imageUUID := *i.Metadata.UUID
+			if imageName == expectedImageName && hasCategoryAssigned(i.Metadata, ec) {
+				found = true
+				o.logger.Infof("Deleting image %s with UUID %s", imageName, imageUUID)
+				response, err := o.v3Client.V3.DeleteImage(imageUUID)
+				if err != nil {
+					return err
+				}
+
+				if err = nutanixtypes.WaitForTask(o.v3Client.V3, response.Status.ExecutionContext.TaskUUID.(string)); err != nil {
+					return err
+				}
+
+				break
+			}
+		}
+
+		if !found {
+			o.logger.Infof("No image with name %s was found", expectedImageName)
+		}
+	}
+	return nil
+}
+
+func cleanupCategories(o *clusterUninstaller) error {
+	ec := createExpectedCategory(o.infraID)
+	_, err := o.v3Client.V3.GetCategoryKey(ec.key)
+	if err != nil {
+		if strings.Contains(err.Error(), "does not exist") {
+			//Already deleted
+			return nil
+		}
+		return err
+	}
+
+	for _, val := range ec.values {
+		o.logger.Infof("Deleting category value : %s", val)
+		err := o.v3Client.V3.DeleteCategoryValue(ec.key, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	o.logger.Infof("Deleting category key : %s", ec.key)
+	return o.v3Client.V3.DeleteCategoryKey(ec.key)
+}
+
+func createExpectedCategory(infraID string) *expectedCategory {
+	return &expectedCategory{
+		key:    fmt.Sprintf(expectedCategoryKeyFormat, infraID),
+		values: []string{expectedCategoryValueOwned, expectedCategoryValueShared},
+	}
+}
+
+func logToBeDeletedVMs(tbd []*nutanixclientv3.VMIntentResource, l logrus.FieldLogger) {
+	l.Info("VMs scheduled to be deleted: ")
+	for _, v := range tbd {
+		l.Infof("- %s", *v.Spec.Name)
+	}
+}
+
+func deleteVMs(clientV3 nutanixclientv3.Service, tbd []*nutanixclientv3.VMIntentResource, l logrus.FieldLogger) error {
+	taskUUIDs := make([]string, 0)
+	for _, v := range tbd {
+		l.Infof("Deleting vm %s with ID %s", *v.Spec.Name, *v.Metadata.UUID)
+		response, err := clientV3.DeleteVM(*v.Metadata.UUID)
+		if err != nil {
+			return err
```

Comment:
I think staebler was looking to have the deletion of VMs to continue even if a failure was encountered.